Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/PropertyEditorAPI_Internals.h

```diff
@@ -354,31 +366,60 @@ namespace AzToolsFramework
+                    {
+                        m_proxyValue = *reinterpret_cast<WrappedType*>(valueToPointerObject->m_address);
+                        typeId = valueToPointerObject->m_typeId;
+                    }
+                    else
+                    {
+                        const AZ::SerializeContext::ClassData* pointerClassData =
+                            serializeContext != nullptr ? serializeContext->FindClassData(valueToPointerObject->m_typeId) : nullptr;
+                        if (pointerClassData != nullptr && pointerClassData->m_azRtti != nullptr)
+                        {
+                            if (auto castInstance = pointerClassData->m_azRtti->Cast<WrappedType>(valueToPointerObject->m_address); castInstance != nullptr)
+                            {
+                                m_proxyValue = *castInstance;
+                                // Set the type ID to the derived class ID
+                                typeId = valueToPointerObject->m_typeId;
+                            }
+                        }
+                    }
+                }
             }
 
-            AZ::SerializeContext* serializeContext = nullptr;
-            AZ::ComponentApplicationBus::BroadcastResult(serializeContext, &AZ::ComponentApplicationBus::Events::GetSerializeContext);
-            AZ_Assert(serializeContext, "Serialization context not available");
-
             // Set the m_genericClassInfo (if any) for our type in case the property handler needs to access it
             // when downcasting to the generic type (e.g. for asset property downcasting to the generic AZ::Data::Asset<AZ::Data::AssetData>)
-            auto typeIdAttribute = node.FindMember(PropertyEditor::ValueType.GetName());
-            AZ::TypeId typeId = AZ::TypeId::CreateNull();
-            if (typeIdAttribute != node.MemberEnd())
-            {
-                typeId = AZ::Dom::Utils::DomValueToTypeId(typeIdAttribute->second);
-            }
-            else if (value.has_value())
-            {
-                typeId = AZ::Dom::Utils::GetValueTypeId(value.value());
-            }
-            if (!typeId.IsNull())
+            if (typeId.IsNull())
             {
-                m_proxyClassElement.m_genericClassInfo = serializeContext->FindGenericClassInfo(typeId);
+                auto typeIdAttribute = node.FindMember(PropertyEditor::ValueType.GetName());
+                if (typeIdAttribute != node.MemberEnd())
+                {
+                    typeId = AZ::Dom::Utils::DomValueToTypeId(typeIdAttribute->second);
+                }
+                else if (value.has_value())
+                {
+                    typeId = AZ::Dom::Utils::GetValueTypeId(value.value());
+                }
+                if (!typeId.IsNull())
+                {
+                    m_proxyClassElement.m_genericClassInfo = serializeContext->FindGenericClassInfo(typeId);
+                }
```

Comment:
I think this should be OK, a quick verification would be to check if setting/clearing a mesh asset still works.

Reply:
Actually this doesn't affect clearing an asset. The logic that handles that case was in the the ReflectionAdapter using JSON Serialization to copy new Dom updated Dom::Value with the Asset reference being cleared to the Dom::Value object representing the instance data.

The AssetJsonSerializer previously would call `AssetJsonSerializer::Store` which works to store an AssetID with guid of all's 0 (a null GUID), but the `AssetJsonSerializer::Load` would check if GUID is null and if so it would clear out the instance being loaded into.
Here is the old code for reference: https://github.com/o3de/o3de/blob/152bc0a1851d881fe735adf54ec93e1ad7875b11/Code/Framework/AzCore/AzCore/Asset/AssetJsonSerializer.cpp#L133-L142

I fixed that issue in this PR https://github.com/o3de/o3de/commit/3393f127db73e410449a06e5daa417786c06c4c2.

Now on the subject of setting the Asset reference, that occurs in a different function `PropertyEditorAPI_Internals.h`.
The [RpePropertyHandlerWrapper::OnValueChanged](https://github.com/o3de/o3de/blob/development/Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/PropertyEditorAPI_Internals.h#L508-L540) function is where a storing convertible type such as an `Asset<MeshAsset>` into an `Asset<AssetData>` occurs.